addons/sale/models/sale.py

```diff
@@ -202,6 +202,13 @@ def _get_payment_type(self):
         ('confirmation_date_conditional_required', "CHECK( (state IN ('sale', 'done') AND confirmation_date IS NOT NULL) OR state NOT IN ('sale', 'done') )", "A confirmed sales order requires a confirmation date."),
     ]
 
+    @api.multi
+    @api.constrains('company_id')
+    def _check_product_from_own_company(self):
+        for line in self.mapped('order_line'):
```

Comment:
`for line in self.mapped('order_line'):` ? 
Maybe it should be `for product_id in self.mapped('product_id'):` ? :

Reply:
note: the usage of `mapped` is no longer necessary to get recrodsets of relational fields
`self.order_line` will return all line from the order in self

Reply:
Correct @dbo-odoo but here we need to check products from all sales order lines so we can use `for product_id in self.mapped('product_id'):`